### ISSUE TITLE
Some questions

### ISSUE THREAD

**Issue by stefandevo** [dj-nitehawk/FastEndpoints#17]
Hi,

Thanks for this great package; trying to see whether useful for our projects.

Some things I am struggling with:

- we have several API (for different apps), but basically they are using a subset of all the logic in our app. How could we make api1 and api2 using/exposing only a subset of the code. Because everything is in one solution, the API (web) is deployed in the same docker container, but we want some endpoints not to be available

- if you have DB specific code that is reused in different end-points, where would you put these?

thx

**Comment by dj-nitehawk** [comment]
you're most welcome!

i'm not exactly sure if i understand your first question. how do you usually do it with asp.net/mvc? however you do, you can do it the exact same way with FastEndpoints i suppose. i don't think there'd be much of a difference in how you'd architect your system. 

are you saying you want to build a monolithic webapi but only want to allow access to some of the endpoints depending on what client it is? if that's the case, you could write a pre-processor like the following in order to allow or deny access to each client/app:

```csharp
public class SecurityPreProcessor<TRequest> : IPreProcessor<TRequest>
{
    public Task PreProcessAsync(TRequest req, HttpContext ctx, List<ValidationFailure> failures, CancellationToken ct)
    {
        var clientType = ctx.Request.Headers["client-type"].FirstOrDefault();

        if (clientType is null)
        {
            failures.Add(new("MissingHeaders", "The [client-type] header needs to be set!"));
            return ctx.Response.SendErrorsAsync(failures);
        }

        if (clientType == "mobile-app" && ctx.Request.Path.StartsWithSegments("api1"))
            return Task.CompletedTask;

        if (clientType == "web-app" && ctx.Request.Path.StartsWithSegments("api2"))
            return Task.CompletedTask;

        return ctx.Response.SendForbiddenAsync();
    }
}
```
basically, every client will send a header value with each request. preprocessor then checks the header value to see if the current client type can access the endpoint/route. routes will obviously start with `/api1/xxxx/yyyy` and `/api2/xxx/yyyy`.

as for shared data access code, you can either do repository pattern and inject repos into your endpoints. or if you're doing vertical slice architecture, you can just put shared db logic in it's own namespace/project/whatever and call those shared methods from the handlers. it's really up to you.

**Comment by stefandevo** [comment]
Hi, thanks for the answers.

Before I had all the logic in separate projects, so I had multiple WebApi projects pointing to the (mediatr) commands in the separate projects. Therefore I could decide in each API (which is deployed with a specific domain name + container) which calls where available. But you are correct, underlying was a monolithic (although using clean architecture) solution.

The PreProcessor could indeed work; the check could come from an environment variable, so I basically have to give the requests attributes for example and check these attributes against the environment variable (instead of the header).

So I should return a NotFoundAsync response for example; would this also work for the Swagger calls? Would I be able to excluded them in the same way?

To have multiple preprocessors for all my requests, I just use a base EndPoint class then to have the processor inside.

**Comment by stefandevo** [comment]
What would help, is that there is a kind of override/func/action in https://github.com/dj-nitehawk/FastEndpoints/blob/3600628efb161f42a706dd939c2c7b0fd58dc096/Src/Library/Extensions/MainExtensions.cs#L38 - that way you could return a false for example if you get an attribute yes/no to "remove" the endpoint from the application

**Comment by stefandevo** [comment]
And the ability to register a global Pre/Post processor?

**Comment by dj-nitehawk** [comment]
> To have multiple preprocessors for all my requests, I just use a base EndPoint class then to have the processor inside.

yeah if you use a preprocessor, that means you'd have to set it on each endpoint. so i think if it's a global thing, it would be a better option to write it as an [asp.net middleware](https://www.tutorialsteacher.com/core/aspnet-core-middleware) which would apply that logic globally to every incoming request.

> the check could come from an environment variable, so I basically have to give the requests attributes for example and check these attributes against the environment variable (instead of the header).

choice is yours. user agent/header/field whatever you like.

> would this also work for the Swagger calls

i think i saw somewhere swagger has an option to hide/show stuff by tag or domain name. unfortunately you'd have to do some research on how to achieve that with swagger. if i remember correctly you'd have to write and register an `OperationProcessor` for nwag or `OperationFilter` for swashbuckle.

**Comment by stefandevo** [comment]
> > To have multiple preprocessors for all my requests, I just use a base EndPoint class then to have the processor inside.
> 
> yeah if you use a preprocessor, that means you'd have to set it on each endpoint. so i think if it's a global thing, it would be a better option to write it as an [asp.net middleware](https://www.tutorialsteacher.com/core/aspnet-core-middleware) which would apply that logic globally to every incoming request.

Ah, of course, I can still do that as well :-)

**Comment by stefandevo** [comment]
> What would help, is that there is a kind of override/func/action in
> 
> https://github.com/dj-nitehawk/FastEndpoints/blob/3600628efb161f42a706dd939c2c7b0fd58dc096/Src/Library/Extensions/MainExtensions.cs#L38
> 
> - that way you could return a false for example if you get an attribute yes/no to "remove" the endpoint from the application

Would this be possible?

**Comment by dj-nitehawk** [comment]
> to "remove" the endpoint from the application

so, to clarify? you are putting all of your endpoints in one monolith. but you want to deploy only a subset of those to seperate servers? is that right?

and having the option to not register certain endpoints found in the app in a given deployment is what you want?

**Comment by stefandevo** [comment]
> > to "remove" the endpoint from the application
> 
> so, to clarify? you are putting all of your endpoints in one monolith. but you want to deploy only a subset of those to seperate servers? is that right?
> 
> and having the option to not register certain endpoints found in the app in a given deployment is what you want?

That's correct; we want to keep our core code in one place, but want to share functionalities over different API projects basically (deployed different servers/nodes). So currency we have different Web API projects on top of our core solution, and mapping these functions that are needed in each Web API. 

Would be great if we just could return a true/false for each "found" endpoint; in this code we can decide to register it or not.

**Comment by dj-nitehawk** [comment]
for example:

```csharp
UseFastEndpoints(exclude: "/api1")
```
something like that is what you want? any endpoint starting with /api1 shouldn't be auto registered?

**Comment by stefandevo** [comment]
> for example:
> 
> ```cs
> UseFastEndpoints(exclude: "/api1")
> ```
> 
> something like that is what you want? any endpoint starting with /api1 shouldn't be auto registered?

I think we would need more control; it's not just the prefix. It could be /client/create in API1 and API2 but not in API3. There is /client/list for example. So basically if we could get a Func in that method to return true/false based upon the Request object would be great.

**Comment by dj-nitehawk** [comment]
k got it... let me look in to that and get back to ya...
probably tomorrow though.

**Comment by stefandevo** [comment]
Sure! Thanks.

**Comment by dj-nitehawk** [comment]
v2.21.0-beta2 should be available on nuget in a few mins.
you can do the following now.
```csharp
app.UseFastEndpoints(ep => ep.Routes.Contains("/api/test") ? true : false);
``
return `true` from the func to exlude an endpoint from registration.
let me know if you need access to any more info in that func.

off to bed. i'll have a look at your other issue in the morning.

**Comment by stefandevo** [comment]
Okay thanks for this.

I just checked; it would be great if you could pass the Type itself also; `FoundEndpoint` for example. That way I could get the attributes for example of the type.

btw: I just found a Reddit Post where you say you build this inspired of ServiceStack. Great, I was also a SS user before and trying to cut out everything I don't need; this lib is great!

**Comment by dj-nitehawk** [comment]
added more in `v2.21.0-beta3`

https://github.com/dj-nitehawk/FastEndpoints/blob/42c895a8ae66c509a5ab5a378936d810f56c9fdc/Src/Library/Extensions/MainExtensions.cs#L220-L230

**Comment by stefandevo** [comment]
Perfect!
